libp2p/muxers/mplex/coder.nim

```diff
@@ -83,4 +83,8 @@ proc writeMsg*(conn: Connection,
                id: uint64,
                msgType: MessageType,
                data: string) {.async, gcsafe.} =
+  # TODO: changing this to
+  #`await conn.writeMsg(id, msgType, cast[seq[byte]](data))`
+  # causes all sorts of race conditions and hangs.
+  # DON'T DO IT!
```

Comment:
If so remove `async`

Reply:
I believe this will lead to swallowed/lost exceptions, but not sure. I really need to take a look at the expanded macro for all this places.

Reply:
Just tried it out locally, by just returning a `Future[void]` and it breaks immediately. In theory, it should behave identically, but of course since it expands to an async iter, something does change.

Reply:
@cheatfate can you check this too, please

Reply:
Here the expanded macros for each:

## Using `result`

```nim
proc writeMsg*(conn: Connection,
               id: uint64,
               msgType: MessageType,
               data: string) {.async, gcsafe.} =
  result = conn.writeMsg(id, msgType, cast[seq[byte]](data))
```
expanded:
```nim
proc writeMsg*(conn: Connection; id: uint64; msgType: MessageType;
              data: seq[byte] = @[]): Future[void] {.gcsafe, stackTrace: off.} =
  var chronosInternalRetFuture = newFuture[void]("writeMsg")
  iterator writeMsg_62640073(): FutureBase {.closure, gcsafe.} =
    trace "sending data over mplex", id, msgType, data = data.len
    try:
      var
        left = data.len
        offset = 0
      while left > 0 or data.len == 0:
        let
          chunkSize = if left > MaxMsgSize:
            MaxMsgSize - 64 else:
            left
          chunk = if chunkSize > 0:
            data[offset .. (offset + chunkSize - 1)] else:
            data
        ## write lenght prefixed
        var buf = initVBuffer()
        buf.writePBVarint(id shl 3 or ord(msgType).uint64)
        buf.writePBVarint(chunkSize.uint64)
        buf.finish()
        left = left - chunkSize
        offset = offset + chunkSize
        await conn.write(buf.buffer & chunk)
        if data.len == 0:
          complete(chronosInternalRetFuture)
          return nil
    except LPStreamEOFError:
      trace "Ignoring EOF while writing"
    except CancelledError as exc:
      raise exc
    except CatchableError as exc:
      debug "Could not write to connection", msg = exc.msg
    complete(chronosInternalRetFuture)

  var nameIterVar`gensym62640075 = writeMsg_62640073
  {.push, stackTrace: off.}
  proc writeMsg_continue_62640074(udata`gensym62640076: pointer = nil) {.closure,
      gcsafe.} =
    try:
      if not (nameIterVar`gensym62640075.finished()):
        var next`gensym62640077 = nameIterVar`gensym62640075()
        while (not next`gensym62640077.isNil()) and
            next`gensym62640077.finished():
          next`gensym62640077 = nameIterVar`gensym62640075()
          if nameIterVar`gensym62640075.finished():
            break
        if next`gensym62640077 == nil:
          if not (chronosInternalRetFuture.finished()):
            let msg`gensym62640078 = "Async procedure ($1) yielded `nil`, " &
                "are you await\'ing a `nil` Future?"
            raise newException(AssertionError, msg`gensym62640078 % "writeMsg")
        else:
          {.gcsafe.}:
            next`gensym62640077.addCallback(writeMsg_continue_62640074)
    except CancelledError:
      chronosInternalRetFuture.cancel()
    except CatchableError as exc`gensym62640079:
      if chronosInternalRetFuture.finished():
        raise exc`gensym62640079
      else:
        chronosInternalRetFuture.fail(exc`gensym62640079)

  writeMsg_continue_62640074()
  {.pop.}
  return chronosInternalRetFuture

proc writeMsg*(conn: Connection; id: uint64; msgType: MessageType; data: string): Future[
    void] {.gcsafe, stackTrace: off.} =
  var chronosInternalRetFuture = newFuture[void]("writeMsg")
  iterator writeMsg_62870044(): FutureBase {.closure, gcsafe.} =
    result = conn.writeMsg(id, msgType, cast[seq[byte]](data))
    complete(chronosInternalRetFuture)

  var nameIterVar`gensym62870046 = writeMsg_62870044
  {.push, stackTrace: off.}
  proc writeMsg_continue_62870045(udata`gensym62870047: pointer = nil) {.closure,
      gcsafe.} =
    try:
      if not (nameIterVar`gensym62870046.finished()):
        var next`gensym62870048 = nameIterVar`gensym62870046()
        while (not next`gensym62870048.isNil()) and
            next`gensym62870048.finished():
          next`gensym62870048 = nameIterVar`gensym62870046()
          if nameIterVar`gensym62870046.finished():
            break
        if next`gensym62870048 == nil:
          if not (chronosInternalRetFuture.finished()):
            let msg`gensym62870049 = "Async procedure ($1) yielded `nil`, " &
                "are you await\'ing a `nil` Future?"
            raise newException(AssertionError, msg`gensym62870049 % "writeMsg")
        else:
          {.gcsafe.}:
            next`gensym62870048.addCallback(writeMsg_continue_62870045)
    except CancelledError:
      chronosInternalRetFuture.cancel()
    except CatchableError as exc`gensym62870050:
      if chronosInternalRetFuture.finished():
        raise exc`gensym62870050
      else:
        chronosInternalRetFuture.fail(exc`gensym62870050)

  writeMsg_continue_62870045()
  {.pop.}
  return chronosInternalRetFuture
```

## Using `await`

```nim
proc writeMsg*(conn: Connection,
               id: uint64,
               msgType: MessageType,
               data: string) {.async, gcsafe.} =
  await conn.writeMsg(id, msgType, cast[seq[byte]](data))
```

expanded:

```nim
proc writeMsg*(conn: Connection; id: uint64; msgType: MessageType;
              data: seq[byte] = @[]): Future[void] {.gcsafe, stackTrace: off.} =
  var chronosInternalRetFuture = newFuture[void]("writeMsg")
  iterator writeMsg_62640073(): FutureBase {.closure, gcsafe.} =
    trace "sending data over mplex", id, msgType, data = data.len
    try:
      var
        left = data.len
        offset = 0
      while left > 0 or data.len == 0:
        let
          chunkSize = if left > MaxMsgSize:
            MaxMsgSize - 64 else:
            left
          chunk = if chunkSize > 0:
            data[offset .. (offset + chunkSize - 1)] else:
            data
        ## write lenght prefixed
        var buf = initVBuffer()
        buf.writePBVarint(id shl 3 or ord(msgType).uint64)
        buf.writePBVarint(chunkSize.uint64)
        buf.finish()
        left = left - chunkSize
        offset = offset + chunkSize
        await conn.write(buf.buffer & chunk)
        if data.len == 0:
          complete(chronosInternalRetFuture)
          return nil
    except LPStreamEOFError:
      trace "Ignoring EOF while writing"
    except CancelledError as exc:
      raise exc
    except CatchableError as exc:
      debug "Could not write to connection", msg = exc.msg
    complete(chronosInternalRetFuture)

  var nameIterVar`gensym62640075 = writeMsg_62640073
  {.push, stackTrace: off.}
  proc writeMsg_continue_62640074(udata`gensym62640076: pointer = nil) {.closure,
      gcsafe.} =
    try:
      if not (nameIterVar`gensym62640075.finished()):
        var next`gensym62640077 = nameIterVar`gensym62640075()
        while (not next`gensym62640077.isNil()) and
            next`gensym62640077.finished():
          next`gensym62640077 = nameIterVar`gensym62640075()
          if nameIterVar`gensym62640075.finished():
            break
        if next`gensym62640077 == nil:
          if not (chronosInternalRetFuture.finished()):
            let msg`gensym62640078 = "Async procedure ($1) yielded `nil`, " &
                "are you await\'ing a `nil` Future?"
            raise newException(AssertionError, msg`gensym62640078 % "writeMsg")
        else:
          {.gcsafe.}:
            next`gensym62640077.addCallback(writeMsg_continue_62640074)
    except CancelledError:
      chronosInternalRetFuture.cancel()
    except CatchableError as exc`gensym62640079:
      if chronosInternalRetFuture.finished():
        raise exc`gensym62640079
      else:
        chronosInternalRetFuture.fail(exc`gensym62640079)

  writeMsg_continue_62640074()
  {.pop.}
  return chronosInternalRetFuture

proc writeMsg*(conn: Connection; id: uint64; msgType: MessageType; data: string): Future[
    void] {.gcsafe, stackTrace: off.} =
  var chronosInternalRetFuture = newFuture[void]("writeMsg")
  iterator writeMsg_62870044(): FutureBase {.closure, gcsafe.} =
    await conn.writeMsg(id, msgType, cast[seq[byte]](data))
    complete(chronosInternalRetFuture)

  var nameIterVar`gensym62870046 = writeMsg_62870044
  {.push, stackTrace: off.}
  proc writeMsg_continue_62870045(udata`gensym62870047: pointer = nil) {.closure,
      gcsafe.} =
    try:
      if not (nameIterVar`gensym62870046.finished()):
        var next`gensym62870048 = nameIterVar`gensym62870046()
        while (not next`gensym62870048.isNil()) and
            next`gensym62870048.finished():
          next`gensym62870048 = nameIterVar`gensym62870046()
          if nameIterVar`gensym62870046.finished():
            break
        if next`gensym62870048 == nil:
          if not (chronosInternalRetFuture.finished()):
            let msg`gensym62870049 = "Async procedure ($1) yielded `nil`, " &
                "are you await\'ing a `nil` Future?"
            raise newException(AssertionError, msg`gensym62870049 % "writeMsg")
        else:
          {.gcsafe.}:
            next`gensym62870048.addCallback(writeMsg_continue_62870045)
    except CancelledError:
      chronosInternalRetFuture.cancel()
    except CatchableError as exc`gensym62870050:
      if chronosInternalRetFuture.finished():
        raise exc`gensym62870050
      else:
        chronosInternalRetFuture.fail(exc`gensym62870050)

  writeMsg_continue_62870045()
  {.pop.}
  return chronosInternalRetFuture
```

## Using `Future`

```nim
proc writeMsg*(conn: Connection,
               id: uint64,
               msgType: MessageType,
               data: string): Future[void] {.gcsafe.} =
  conn.writeMsg(id, msgType, cast[seq[byte]](data))
```
expanded:

```nim
proc writeMsg*(conn: Connection; id: uint64; msgType: MessageType;
              data: seq[byte] = @[]): Future[void] {.gcsafe, stackTrace: off.} =
  var chronosInternalRetFuture = newFuture[void]("writeMsg")
  iterator writeMsg_62640073(): FutureBase {.closure, gcsafe.} =
    trace "sending data over mplex", id, msgType, data = data.len
    try:
      var
        left = data.len
        offset = 0
      while left > 0 or data.len == 0:
        let
          chunkSize = if left > MaxMsgSize:
            MaxMsgSize - 64 else:
            left
          chunk = if chunkSize > 0:
            data[offset .. (offset + chunkSize - 1)] else:
            data
        ## write lenght prefixed
        var buf = initVBuffer()
        buf.writePBVarint(id shl 3 or ord(msgType).uint64)
        buf.writePBVarint(chunkSize.uint64)
        buf.finish()
        left = left - chunkSize
        offset = offset + chunkSize
        await conn.write(buf.buffer & chunk)
        if data.len == 0:
          complete(chronosInternalRetFuture)
          return nil
    except LPStreamEOFError:
      trace "Ignoring EOF while writing"
    except CancelledError as exc:
      raise exc
    except CatchableError as exc:
      debug "Could not write to connection", msg = exc.msg
    complete(chronosInternalRetFuture)

  var nameIterVar`gensym62640075 = writeMsg_62640073
  {.push, stackTrace: off.}
  proc writeMsg_continue_62640074(udata`gensym62640076: pointer = nil) {.closure,
      gcsafe.} =
    try:
      if not (nameIterVar`gensym62640075.finished()):
        var next`gensym62640077 = nameIterVar`gensym62640075()
        while (not next`gensym62640077.isNil()) and
            next`gensym62640077.finished():
          next`gensym62640077 = nameIterVar`gensym62640075()
          if nameIterVar`gensym62640075.finished():
            break
        if next`gensym62640077 == nil:
          if not (chronosInternalRetFuture.finished()):
            let msg`gensym62640078 = "Async procedure ($1) yielded `nil`, " &
                "are you await\'ing a `nil` Future?"
            raise newException(AssertionError, msg`gensym62640078 % "writeMsg")
        else:
          {.gcsafe.}:
            next`gensym62640077.addCallback(writeMsg_continue_62640074)
    except CancelledError:
      chronosInternalRetFuture.cancel()
    except CatchableError as exc`gensym62640079:
      if chronosInternalRetFuture.finished():
        raise exc`gensym62640079
      else:
        chronosInternalRetFuture.fail(exc`gensym62640079)

  writeMsg_continue_62640074()
  {.pop.}
  return chronosInternalRetFuture
```

Reply:
ping @cheatfate - this looks fishy

Reply:
If you want to avoid race conditions do not use `await` and declare your procedure without `async` pragma, in such way it will be just a simple type conversion proxy. As soon as you introduce `await` you will get race conditions. Also you can protect your `Connection` resource with AsyncLock, in such way you can avoid race conditions and it will also help you to maintain proper order (first started - first served).

Reply:
The inherent race conditions are understood, what's weird is the unexpected behavior from using `result`. It shouldn't even compile but it does.

Reply:
Actually, it makes sense that it compiles, in the expanded macro it's clear that the wrapped `writeMsg` return a future, which matches the return value of the called `writeMsg`. What's still unexpected is the change in behaviour, but your explanations about the additional indirection with await makes sense and matches my own suspicion. I guess this isn't necessarily broken, but a bit of a gotcha to keep in mind always.

Reply:
We should review our entire codebase and see if there is any other places where this can/does happen.

Reply:
I will check it better but I don't understand the last, what does expanded mean for the last ? There is no macro to expand...

Reply:
@cheatfate how come case 3 is expanding? and what macro?

Reply:
@sinkingsugar you are right case 3 can't be expanded of course.
All the expansions @dryajov posted is not expansions of 
```nim
proc writeMsg*(conn: Connection, id: uint64, msgType: MessageType, data: string)
```
Because as you can see expansion here should provide last argument `data` type as `string`, but all expansions made for procedure with last argument `data` type as `seq[byte]`. So its why all procedures has expansions.

Reply:
If you look at the expansions, both result and await have two methods, future has one as @cheatfate pointed out, so the last one didn't expand actually.